tests/CPT_test.py

```diff
@@ -0,0 +1,332 @@
+from typing import Any, Dict, List, Union
+
+import pytest
+import torch
+from datasets import load_dataset
+from torch.utils.data import Dataset
+from tqdm import tqdm
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    DataCollatorForLanguageModeling,
+    Trainer,
+    TrainingArguments,
+)
+
+from peft import CPTConfig, get_peft_model
+
+
+TEMPLATE = {"input": "input: {}", "intra_seperator": " ", "output": "output: {}", "inter_seperator": "\n"}
+
+MODEL_NAME = "bigscience/bloom-1b7"
+MAX_INPUT_LENGTH = 1024
+
+
+@pytest.fixture(scope="module")
+def global_tokenizer():
+    """Load the tokenizer fixture for the model."""
+
+    return AutoTokenizer.from_pretrained(MODEL_NAME, cache_dir=".", padding_side="right", trust_remote_code=True)
+
+
+@pytest.fixture(scope="module")
+def config_text():
+    """Load the SST2 dataset and prepare it for testing."""
+    config = CPTConfig(
+        CPT_token_ids=[0, 1, 2, 3, 4, 5, 6, 7],  # Example token IDs for testing
+        CPT_mask=[1, 1, 1, 1, 1, 1, 1, 1],
+        CPT_tokens_type_mask=[1, 2, 2, 2, 3, 3, 3, 4],
+        CPT_prompt_tuning_init="TEXT",
+        num_virtual_tokens=8,
+        opt_weighted_loss_type="decay",
+        opt_loss_decay_factor=0.95,
+        opt_projection_epsilon=0.2,
+        opt_projection_format_epsilon=0.1,
+        tokenizer_name_or_path=MODEL_NAME,
+    )
+    return config
+
+
+@pytest.fixture(scope="module")
+def config_random():
+    """Load the SST2 dataset and prepare it for testing."""
+    config = CPTConfig(
+        CPT_prompt_tuning_init="RANDOM",
+        num_virtual_tokens=8,
+        opt_weighted_loss_type="decay",
+        opt_loss_decay_factor=0.95,
+        opt_projection_epsilon=0.2,
+        opt_projection_format_epsilon=0.1,
+        tokenizer_name_or_path=MODEL_NAME,
+    )
+    return config
+
+
+@pytest.fixture(scope="module")
+def sst_data():
+    """Load the SST2 dataset and prepare it for testing."""
+    data = load_dataset("glue", "sst2")
+
+    def add_string_labels(example):
+        if example["label"] == 0:
+            example["label_text"] = "negative"
+        elif example["label"] == 1:
+            example["label_text"] = "positive"
+        return example
+
+    train_dataset = data["train"].select(range(4)).map(add_string_labels)
+    test_dataset = data["validation"].select(range(10)).map(add_string_labels)
```

Comment:
We only use a small sample of the whole dataset, right? Instead of loading the whole dataset and filtering it down so much, how about just hard-coding these few samples? Alternatively, we could create a mini dataset and load that instead.